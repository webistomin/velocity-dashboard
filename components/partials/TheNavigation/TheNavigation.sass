+b('main-nav')
  width: 215px
  height: calc(100% - 5.1rem)
  position: fixed
  top: 5.1rem
  left: 0
  z-index: var(--z-index-navigation)
  visibility: hidden
  opacity: 0
  transition: opacity 0.2s ease-out, visibility 0.2s ease-out

  +m('visible')
    visibility: visible
    opacity: 1

    .main-nav__wrapper
      transition-delay: 0.2s
      transform: none

  +media($screen-sm)
    top: 8rem
    height: calc(100% - 8rem)

  +media($screen-lg)
    width: 80px
    height: 100%
    grid-area: navigation
    position: relative
    top: auto
    left: auto
    background-color: rgb(var(--color-white))
    border-right: 1px solid rgba(var(--color-primary), 0.2)
    overflow: initial
    visibility: visible
    opacity: 1
    transition: width 0.2s ease-out
    will-change: width
    backface-visibility: hidden

    +m('visible')
      width: 215px

  +e('overlay')
    z-index: -1
    height: inherit
    top: inherit

    +media($screen-lg)
      display: none

  +e('wrapper')
    position: relative
    height: 100%
    overflow: auto
    +scroll-overflow()
    transform: translateX(-100%)
    background-color: rgb(var(--color-white))
    transition: transform 0.2s ease-out

    +media($screen-lg)
      top: 0
      position: sticky
      transform: none
      height: 100vh
      overflow-y: auto
      overflow-x: hidden
      transition: none
      +scrollbar(5px, var(--color-primary), var(--color-secondary))

  +e('toggle')
    display: none

    +media($screen-xl)
      display: block

  +e('user')
    display: flex
    align-items: center
    padding: rem(14px) rem(14px) rem(28px)

    +media($screen-sm)
      padding-left: rem(24px)
      padding-right: rem(24px)

    +media($screen-lg)
      padding-top: rem(22px)
      padding-bottom: rem(22px)

    +media($screen-xl)
      padding-top: rem(30px)
      padding-bottom: rem(30px)

  +e('info')
    display: flex
    flex-direction: column
    font: var(--text-link)
    transition: opacity 0.2s ease-out, visibility 0.2s ease-out

    +media($screen-lg)
      opacity: 0
      visibility: hidden

      +m('visible')
        opacity: 1
        visibility: visible

  +e('thumb')
    margin-right: rem(10px)

    +media($screen-lg)
      flex-shrink: 0

  +e('welcome')
    font: inherit

  +e('username')
    font: inherit
    color: rgb(var(--color-gray-3))
    +text-overflow()
    max-width: 12ch

  +e('link')
    display: flex
    align-items: center
    padding: rem(16px) rem(14px)
    color: rgb(var(--color-gray-3))
    transition: color 0.2s ease-out, background-color 0.2s ease-out
    position: relative

    .icon
      margin-right: rem(16px)
      flex-shrink: 0

    &::before
      content: ''
      background-color: currentColor
      position: absolute
      top: 0
      left: 0
      width: rem(3px)
      height: 100%
      opacity: 0
      transition: opacity 0.2s ease-out

    &::after
      content: ''
      width: 100%
      height: 100%
      position: absolute
      top: 0
      left: 0
      z-index: 0
      background-color: rgb(var(--color-primary))
      opacity: 0
      transition: opacity 0.2s ease-out, background-color 0.2s ease-out

    &:hover,
    &.nuxt-link-active
      color: rgb(var(--color-primary))

      &::before
        opacity: 1

      &::after
        opacity: 0.15

    +media($screen-sm)
      padding-left: rem(24px)
      padding-right: rem(24px)

    +media($screen-lg)
      padding-left: rem(28px)
      padding-right: rem(28px)

  +e('link-name')
    +media($screen-lg)
      opacity: 0
      visibility: hidden
      transition: opacity 0.2s ease-out, visibility 0.2s ease-out

      +m('visible')
        opacity: 1
        visibility: visible
