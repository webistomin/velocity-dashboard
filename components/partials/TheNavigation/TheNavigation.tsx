import { VueComponent } from 'types/vue-components';
import { Component, Emit, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseToggle from 'components/base/BaseToggle';
import BaseThumbnail from 'components/base/BaseThumbnail';
import BaseOverlay from 'components/base/BaseOverlay';

import './TheNavigation.sass';

export interface INavigationProps {
  isNavOpened: boolean;
  onOpenNav?: () => void;
}

export interface INavigationLinks {
  name: string;
  icon: string;
  to: string;
}

@Component({
  name: 'TheNavigation',
})
export default class TheNavigation extends VueComponent<INavigationProps> {
  @Prop({ default: false })
  private readonly isNavOpened!: INavigationProps['isNavOpened'];

  public navLinks: INavigationLinks[] = [
    {
      name: 'Overview',
      icon: 'icon-grid',
      to: '/',
    },
    {
      name: 'Analytics',
      icon: 'icon-histograms',
      to: '/analytics',
    },
    {
      name: 'Vehicles',
      icon: 'icon-car',
      to: '/vehicles',
    },
    {
      name: 'Service',
      icon: 'icon-service',
      to: '/reminders',
    },
    {
      name: 'Map',
      icon: 'icon-map',
      to: '/map',
    },
    {
      name: 'Chat',
      icon: 'icon-mail',
      to: '/chat',
    },
    {
      name: 'Settings',
      icon: 'icon-settings',
      to: '/settings',
    },
  ];

  @Emit('openNav')
  public onToggleClick(): void {}

  public render(): VNode {
    return (
      <aside class={`main-nav ${this.isNavOpened ? 'main-nav_visible' : ''}`}>
        <div class='main-nav__wrapper'>
          <BaseToggle class='main-nav__toggle' size='s' onClick={this.onToggleClick} isActive={this.isNavOpened} />
          <div class='main-nav__user'>
            <BaseThumbnail class='main-nav__thumb' image='/img/avatar.png' size='xs' alt='user' isSquared={true} />
            <span class={`main-nav__info ${this.isNavOpened ? 'main-nav__info_visible' : ''}`}>
              <strong class='main-nav__welcome'>Welcome</strong>
              <strong class='main-nav__username'>Alexey Istomin</strong>
            </span>
          </div>
          <nav class='main-nav__nav'>
            <ul class='main-nav__list list'>
              {this.navLinks.map((link) => {
                return (
                  <li class='main-nav__item list-item' key={link.name}>
                    <nuxt-link class='main-nav__link link' to={link.to}>
                      <svg-icon name={link.icon} width={24} height={24} />
                      <span class={`main-nav__link-name ${this.isNavOpened ? 'main-nav__link-name_visible' : ''}`}>
                        {link.name}
                      </span>
                    </nuxt-link>
                  </li>
                );
              })}
            </ul>
          </nav>
        </div>
        <BaseOverlay onClick={this.onToggleClick} isVisible={this.isNavOpened} class='main-nav__overlay' />
      </aside>
    );
  }
}
