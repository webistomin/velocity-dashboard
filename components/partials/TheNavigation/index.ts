import TheNavigation from './TheNavigation';

export default TheNavigation;
