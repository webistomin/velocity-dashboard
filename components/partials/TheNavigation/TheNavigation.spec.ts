import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import TheNavigation from './TheNavigation';

describe('TheNavigation', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  let store: any;
  let state;

  beforeEach(() => {
    state = {
      user: {
        role: 'admin',
      },
    };
    store = new Vuex.Store({
      modules: {
        auth: {
          state,
        },
      },
    });
  });

  const factory = (propsData = {}, options = {}) => {
    return shallowMount(TheNavigation, {
      localVue,
      store,
      propsData: {
        isNavOpened: false,
        ...propsData,
      },
      mocks: {
        $auth: {
          user: {
            role: 'admin',
          },
        },
      },
      stubs: ['svg-icon', 'nuxt-link'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Emit "openNav" event', () => {
    const wrapper = factory();
    wrapper.vm.onToggleClick();
    expect(wrapper.emitted().openNav);
  });

  it('Set class from isVisible prop', () => {
    const expectedResult = ['main-nav_visible'];
    const wrapper = factory({
      isNavOpened: true,
    });
    const classes = wrapper.classes();
    expect(classes).toEqual(expect.arrayContaining(expectedResult));
  });

  it('Set isAdmin to false if user role is not admin', () => {
    const wrapper = factory(
      {},
      {
        mocks: {
          $auth: {},
        },
      }
    );
    const isAdmin = wrapper.vm.isUserAdmin;
    expect(isAdmin).toEqual(false);
  });
});
