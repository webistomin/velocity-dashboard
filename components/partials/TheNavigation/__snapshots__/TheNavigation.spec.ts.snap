// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheNavigation Matches snapshot 1`] = `
"<aside class=\\"main-nav \\">
  <div class=\\"main-nav__wrapper\\">
    <basetoggle-stub size=\\"s\\" class=\\"main-nav__toggle\\"></basetoggle-stub>
    <div class=\\"main-nav__user\\">
      <vuecomponent-stub image=\\"\\" alt=\\"user\\" size=\\"xs\\" issquared=\\"true\\" class=\\"main-nav__thumb\\"></vuecomponent-stub><span class=\\"main-nav__info \\"><strong class=\\"main-nav__welcome\\">Welcome</strong><strong class=\\"main-nav__username\\">Alexey Istomin</strong></span>
    </div>
    <nav class=\\"main-nav__nav\\">
      <ul class=\\"main-nav__list list\\">
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/\\" aria-label=\\"Overview\\" title=\\"Overview\\" exact=\\"true\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-grid\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Overview</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/analytics\\" aria-label=\\"Analytics\\" title=\\"Analytics\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-histograms\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Analytics</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/vehicles\\" aria-label=\\"Vehicles\\" title=\\"Vehicles\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-car\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Vehicles</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/reminders\\" aria-label=\\"Service\\" title=\\"Service\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-service\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Service</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/map\\" aria-label=\\"Map\\" title=\\"Map\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-map\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Map</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/chat\\" aria-label=\\"Chat\\" title=\\"Chat\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-mail\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Chat</span>
          </nuxt-link-stub>
        </li>
        <li class=\\"main-nav__item list-item\\">
          <nuxt-link-stub to=\\"/settings\\" aria-label=\\"Settings\\" title=\\"Settings\\" class=\\"main-nav__link link\\">
            <svg-icon-stub name=\\"icon-settings\\" width=\\"24\\" height=\\"24\\"></svg-icon-stub><span class=\\"main-nav__link-name \\">Settings</span>
          </nuxt-link-stub>
        </li>
      </ul>
    </nav>
  </div>
  <baseoverlay-stub class=\\"main-nav__overlay\\"></baseoverlay-stub>
</aside>"
`;
